### PR TITLE
Support `@sparse` constrained map shapes and list shapes

### PR DESCRIPTION
**Pull request by david-perez**
Turns out we've never supported them, neither directly constrained nor
with constrained members, because of a lack of tests. Yet another data
point to prioritize working on code-generating `constraints.smithy` (see
https://github.com/awslabs/smithy-rs/issues/2101).

The implementation is simple: we just need to call the symbol provider
on the member symbols instead of on the target symbols so we get
`Option<T>` list members / map values if applicable, and handle the
wrapper when converting between unconstrained and constrained types with
help from `match` and `Option<T>::map`.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt [hunk 1]
```diff
@@ -102,31 +105,57 @@ class UnconstrainedMapGenerator(
...
+                                    // DRYing this up with the else branch below would make this less understandable.
+                                    rustTemplate(
+                                        """
+                                        match #{ConstrainedValueSymbol}::try_from(v) {
+                                            Ok(v) => Ok((k, Some(v))),
+                                            Err(inner_constraint_violation) => Err(Self::Error::Value(k, inner_constraint_violation)),
+                                        }
+                                        """,
+                                        "ConstrainedValueSymbol" to constrainedValueSymbol,
+                                        "Epilogue" to epilogueWritable,
```
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt:142]:
```suggestion
```

Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt:142]: This doesn't look to be used here in this block

**Review by crisidev — APPROVED** [review]
Just a small comment on a value that looks to be unused. LGTM.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt (hunk 1 above)
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/UnconstrainedMapGenerator.kt]: Fixed in https://github.com/awslabs/smithy-rs/pull/2213/commits/b4ae60f85991e647748c2bc1a1bce247761ca1f3.

**Review by crisidev — APPROVED** [review]

**Pull request merged**